x-pack/plugins/security_solution/public/detections/components/rules/step_define_rule/index.tsx

```diff
@@ -315,18 +398,179 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
     ),
     [
       handleResetThreatIndices,
-      indexPatterns,
+      indexPattern,
       threatBrowserFields,
       threatIndexModified,
       threatIndexPatterns,
       threatIndexPatternsLoading,
     ]
   );
+
+  const dataViewIndexPatternToggleButtonOptions: EuiButtonGroupOptionProps[] = useMemo(
+    () => [
+      {
+        id: INDEX_PATTERN_SELECT_ID,
+        label: i18nCore.translate(
+          'xpack.securitySolution.ruleDefine.indexTypeSelect.indexPattern',
+          {
+            defaultMessage: 'Index Patterns',
+          }
+        ),
+        iconType:
+          dataSourceRadioIdSelected === INDEX_PATTERN_SELECT_ID ? 'checkInCircleFilled' : 'empty',
+        'data-test-subj': `rule-index-toggle-${INDEX_PATTERN_SELECT_ID}`,
+      },
+      {
+        id: DATA_VIEW_SELECT_ID,
+        label: i18nCore.translate('xpack.securitySolution.ruleDefine.indexTypeSelect.dataView', {
+          defaultMessage: 'Data View',
+        }),
+        iconType:
+          dataSourceRadioIdSelected === DATA_VIEW_SELECT_ID ? 'checkInCircleFilled' : 'empty',
+        'data-test-subj': `rule-index-toggle-${DATA_VIEW_SELECT_ID}`,
+      },
+    ],
+    [dataSourceRadioIdSelected]
+  );
+
+  const DataViewSelectorMemo = useMemo(() => {
+    return (
+      <UseField
+        key="DataViewSelector"
+        path="dataViewId"
+        component={DataViewSelector}
+        componentProps={{
+          kibanaDataViews,
+          setIndexPattern,
+          setIsIndexPatternLoading,
+        }}
+      />
+    );
+  }, [kibanaDataViews]);
+  const DataSource = useMemo(() => {
+    return (
+      <RuleTypeEuiFormRow $isVisible={true} fullWidth>
+        <EuiFlexGroup direction="column">
+          <EuiFlexItem>
+            <StyledFlexGroup direction="row" alignItems="stretch">
+              <EuiFlexItem grow={1}>
+                <StyledButtonGroup
+                  legend="Rule index pattern or data view selector"
+                  idSelected={dataSourceRadioIdSelected}
+                  onChange={onChangeDataSource}
+                  options={dataViewIndexPatternToggleButtonOptions}
+                  color="primary"
+                />
```

Comment:
Perhaps you already tried, but would be nice to extract these out into custom components. Though looking at it things seems pretty dependent on each-other, so probably bit difficult to do at this point. We'll need to refactor these steps and our use of the forms lib to have better encapsulation here.

Reply:
Agreed - would love to reduce the overall size of this component. I think breaking the data source selection out into it's own component would be ideal.